Add Text.Parsing.Parser.Language and Token modules.

cdepillabout opened the pull request:
This commit adds the `Text.Parsing.Parser.Language` and
`Text.Parsing.Parser.Token` modules.  This is a straight port of modules
of the same name in Haskell's `parsec` library.

  - [`Text.Parsec.Language`](https://hackage.haskell.org/package/parsec-3.1.9/docs/Text-Parsec-Language.html)
  - [`Text.Parsec.Token`](https://hackage.haskell.org/package/parsec-3.1.9/docs/Text-Parsec-Token.html)

This commit also adds the `(<??>)` combinator, which acts like a flipped
`(<?>)`.  It is convenient to use with monadic parsers:

```purescript
foo :: Parser String Unit
foo = "some message" <??> do
    string "foo"
    bar
    pure unit
```

cdepillabout commented:
This commit adds [a dependency](https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-0a08a7565aba4405282251491979bb6bR33) on `purescript-unicode`.

If this dependency is too heavy, there are a few ways to get around it:

1. Release this PR as a separate library (just like what happened with purescript-unicode).

2. Functions from `purescript-unicode` are being used in the following places:

    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R759
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R762
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R765
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R768
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R771
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R774
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R777
    - https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R777

  It would be possible to reimplement these functions in the `Text.Parsing.Parser.Token` module just working on ASCII.  The downside of this is that the parsers wouldn't be able to handle some unicode inputs.

cdepillabout commented:
One small problem I couldn't figure out was the functions near the bottom of the `Text.Parsing.Parser.Token` module:

https://github.com/purescript-contrib/purescript-parsing/pull/27/files#diff-13c443ec289a466b24a254814b927fa7R675

I tried to put them in the `where` clause of the `makeTokenParser` function, but I was running into the following two problems:

1. Depending on how I wrote the functions, sometimes the compiler was crashing with an error about the list passed to `foldl1` being empty.
2. Sometimes I was getting a compiler error about not being able to have mutually recursive functions in a where clause.

I moved the functions in question to the top level and the problems disappered.

michaelficarra commented on bower.json:
```diff
@@ -29,7 +29,9 @@
     "purescript-strings": "~0.7.0",
-    "purescript-transformers": "^0.8.1"
+    "purescript-transformers": "^0.8.1",
+    "purescript-unicode": "git://github.com/purescript-contrib/purescript-unicode#unicode",
```
You should publish this to bower. Never actually commit git dependencies, especially when it's not a SHA.

cdepillabout replied:
@michaelficarra Thanks for taking a look at this PR.

I am planning to publish purescript-unicode to bower, but I am waiting for [this PR](https://github.com/purescript-contrib/purescript-unicode/pull/1) to get reviewed.

I don't think I explained it well enough in the PR comments, but this PR shouldn't be merged until purescript-unicode is released.  (Or if we decide not to use the purescript-unicode dependency, then it doesn't matter.)

paf31 commented on src/Text/Parsing/Parser/Token.purs:
```diff
@@ -31,3 +47,732 @@ when tokpos f = try $ do
+-- appropiate language definition and selects the lexical parsers that
+-- are needed from the resulting 'GenTokenParser'.
+--
+-- >  module Main where
```
These could be Markdown code blocks.

paf31 commented on src/Text/Parsing/Parser/Token.purs:
```diff
@@ -31,3 +47,732 @@ when tokpos f = try $ do
+-----------------------------------------------------------
+
+-- | The expression @makeTokenParser language@ creates a 'GenTokenParser'
+-- record that contains lexical parsers that are
```
Need to use a pipe to mark every line as a comment. Also, please use Markdown.

paf31 commented on src/Text/Parsing/Parser/Language.purs:
```diff
@@ -0,0 +1,125 @@
+
+module Text.Parsing.Parser.Language
+    -- ( haskellDef, haskell
```
Let's have explicit exports.

paf31 commented:
What @michaelficarra said, plus a few code comments from me, plus:

- We should update the LICENSE file
- Could you please add some tests which mention `haskellStyle` or `javaStyle` specifically?

Thanks very much for working on this!

cdepillabout commented:
@paf31, Thanks for taking the time to review the code.  I have made the following changes:

- In commit 115989c, I changed the exports in `Text.Parsing.Parser.Language` and `ext.Parsing.Parser.Token` to be explicit.
- In commit 072ef28, I changed the documentation from Haddock-style to Markdown.
- In commit 930e594, I added Parsec's license to the LICENSE file.
- In commit 6c2745c, I added some simple tests for the `haskellStyle` and `javaStyle` LanguageDefs.
- In commit cc562ab, I fixed the purescript-unicode dependency in `bower.json`.

Now this PR should be ready for merging.  (Or at least additional review.)

paf31 commented:
Sorry, I'm catching up on PRs. I'll get to this soon!

cdepillabout commented:
No worries!

paf31 commented:
Thanks very much! And sorry again for the delay.
